ArduPlane/mode.cpp

```diff
@@ -120,3 +120,46 @@ bool Mode::is_vtol_man_throttle() const
 #endif
     return false;
 }
+
+// Deny changing into forward flight modes at low altitude
+bool Mode::allow_enter_at_current_alt() const
+{
+#if HAL_QUADPLANE_ENABLED
+    if (!hal.util->get_soft_armed() ||
+            is_vtol_mode() ||
+            !quadplane.option_set(QuadPlane::OPTION::FW_MODE_CHANGE_ALT_LIMIT)) {
+        // always allowed to change mode if disarmed
+        // always allowed to enter a vtol mode
+        // check disabled
+        return true;
+    }
+    float min_alt_cm = 0.0;
+    bool have_min_alt = false;
+#if AP_FENCE_ENABLED
+    if (((plane.fence.get_enabled_fences() & AC_FENCE_TYPE_ALT_MIN) != 0)) {
+        min_alt_cm = plane.fence.get_safe_alt_min()*100;
+        have_min_alt = true;
+    }
+#endif
+    if (plane.g.FBWB_min_altitude_cm > 0) {
+        min_alt_cm = MAX(min_alt_cm, plane.g.FBWB_min_altitude_cm);
+        have_min_alt = true;
+    }
+    if (!have_min_alt) {
+        // no altitude limit set
+        return true;
+    }
+    int32_t alt_cm;
+    if (!plane.current_loc.initialised() || !plane.current_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, alt_cm)) {
```

Comment:
We could use the height above ground estimate from the rangefinder here. But then were using a different height reference than fence applys to its min altitude.